lightning/src/ln/wire.rs

```diff
@@ -219,22 +231,32 @@ pub fn read<R: io::Read>(buffer: &mut R) -> Result<Message, msgs::DecodeError> {
 /// # Errors
 ///
 /// Returns an I/O error if the write could not be completed.
-pub fn write<M: Encode + Writeable, W: Writer>(message: &M, buffer: &mut W) -> Result<(), io::Error> {
-	M::TYPE.write(buffer)?;
+pub fn write<M: Type + Writeable, W: Writer>(message: &M, buffer: &mut W) -> Result<(), io::Error> {
+	message.type_id().write(buffer)?;
 	message.write(buffer)
 }
 
-/// Defines a type-identified encoding for sending messages over the wire.
+mod encode {
+	/// Defines a constant type identifier for reading messages from the wire.
+	pub trait Encode {
+		/// The type identifying the message payload.
+		const TYPE: u16;
+	}
+}
+
+pub(crate) use self::encode::Encode;
+
+/// Defines a type identifier for sending messages over the wire.
 ///
 /// Messages implementing this trait specify a type and must be [`Writeable`] to use with [`write()`].
-pub trait Encode {
-	/// The type identifying the message payload.
-	const TYPE: u16;
-
-	/// Returns the type identifying the message payload. Convenience method for accessing
-	/// [`Self::TYPE`].
-	fn type_id(&self) -> MessageType {
-		MessageType(Self::TYPE)
+pub trait Type {
+	/// Returns the type identifying the message payload.
+	fn type_id(&self) -> u16;
```

Comment:
Not sure how you would do that? Also this trait is used also for non custom message now so we'd need a different trait I guess? And tbh I'm not convinced it would be a great restriction to have, it feels like it could be more something annoying than really useful, but that's just my opinion.

Reply:
> Not sure how you would do that? Also this trait is used also for non custom message now so we'd need a different trait I guess?

Yeah afaict future spec usage of the range `Setup & Control` / `Channel` / `Commitment` / `Routing`. I think such usages should be covered by the compliant part of LDK, i.e in the named part of our `enum Message`